### PR TITLE
Custom help message for people trying to make macro public

### PR DESCRIPTION
**Pull request by est31**
The current help message is too much about "normal" macros to be used
as general message. Keep it for normal macros, and add custom help and
error messages for macro definitions.

### CODE REVIEW

**Comment by brson** [comment]
I like it. Opinions @nikomatsakis @Manishearth ?

File: src/test/compile-fail/pub-macro-rules.rs [hunk 1]
```diff
@@ -0,0 +1,27 @@
+// Copyright 2015 The Rust Project Developers. See the COPYRIGHT
+// file at the top-level directory of this distribution and at
+// http://rust-lang.org/COPYRIGHT.
+//
+// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
+// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
```
Comment (Manishearth) [inline src/test/compile-fail/pub-macro-rules.rs]: This should be a parse-fail test instead?

**Comment by Manishearth** [comment]
r=me once the test is moved

File: src/libsyntax/parse/parser.rs [hunk 2]
```diff
@@ -4614,9 +4614,18 @@ impl<'a> Parser<'a> {
     fn complain_if_pub_macro(&mut self, visa: Visibility, span: Span) {
         match visa {
             Public => {
-                self.span_err(span, "can't qualify macro invocation with `pub`");
-                self.fileline_help(span, "try adjusting the macro to put `pub` inside \
-                                      the invocation");
+                let is_macro_rules :bool = match self.token {
```
Comment (durka) [inline src/libsyntax/parse/parser.rs]: `: `

File: src/test/compile-fail/pub-macro-rules.rs (hunk 1 above)
Comment (est31) [inline src/test/compile-fail/pub-macro-rules.rs]: I'll move it.
Should I move the files `src/test/compile-fail/pub-item-macro.rs` and `src/test/compile-fail/pub-method-macro.rs` too?

Comment (Manishearth) [inline src/test/compile-fail/pub-macro-rules.rs]: If they deal with errors that come from the parser (looks like they do?), then yes.

**Comment by est31** [comment]
rebased

**Pull request merged**